workflow/controller/exit_handler.go

```diff
@@ -57,7 +57,9 @@ func (woc *wfOperationCtx) resolveExitTmplArgument(args wfv1.Arguments, prefix s
 	}
 	if outputs != nil {
 		for _, param := range outputs.Parameters {
-			scope.addParamToScope(fmt.Sprintf("%s.outputs.parameters.%s", prefix, param.Name), param.Value.String())
+			if param.Value != nil {
+				scope.addParamToScope(fmt.Sprintf("%s.outputs.parameters.%s", prefix, param.Name), param.Value.String())
```

Comment:
what will the error be if the parameter is `nil`? because an `unknown name` error like #13061 may be confusing. I wonder if empty string might be better? not entirely sure

Reply:
While debugging, I noticed that the first crash was caused by the `step-status` not being populated when the `resolveExitHandlerTmpl` was called. But once the status is available, if I recall correctly on the next iteration of operate the parameter is filled, and at the end, we can see it as an input (screenshot at verification with `Failed`)

Basically, I noticed before the fix:

1.  Step is failing.
2.  Exit handler node is not created due to `nil` pointer when resolving the template.
3.  Workflow is marked as done (2 nodes).

After the fix:

1.  Step is failing.
2.  Exit node handler is created and we get the `step-status` value.
3.  Workflow is marked as done only after exit handler node is finished.

Reply:
Yes, that's because you have `step-status` here, I'm wondering if a user actually attempts to use `hello-param` as mention [above](https://github.com/argoproj/argo-workflows/pull/13448#discussion_r1712689839), what kind of error they'll get

Reply:
After testing it will get the same panic. The only difference is that will not resolve the template 
```
level=debug msg="unresolved is allowed " error=unresolved
```
and it will continue with the execution and input as shown below.
<img width="1448" alt="image" src="https://github.com/user-attachments/assets/a119d3fa-b5bd-42cb-a27e-f722a6b8c510">

I am not quite sure which case is better empty string or does not add to the scope cause both will require digging into what happened with the previous step. Maybe an empty string is "safer".

Reply:
Ah the dreaded unresolved check... That's meant to be used to defer a resolution when not possible at the current time (e.g. no output _yet_). But in this case, it will _never_ be possible.

The question is if we can consistently determine that is the case without breaking anything.

If so, yes an empty string would make more sense as the zero/null value of string (since nil itself wouldn't be valid here). A user could also detect that in an expression and handle the empty case then too.

Cc @isubasinghe might be worth a check from you as well given that you've dealt with the unresolved check before too

Reply:
Also some closely related issues on non-existence checks: #11549, #12739, #12317

We'd want to make sure #12739 still works. And this might actually properly fix #12317?

Reply:
Thank you very much for the detailed information. I will wait for the valuable input from @isubasinghe and in the meantime I will try to test/debug the above issues.

Reply:
The first two I linked already have root causes for reference. The last one may not. We do need better testing of expressions in general though as they can be somewhat easy to accidentally introduce breaking changes to. (and have some bugs, as well as inconsistencies in templating languages that I'd planned to eventually unify per #9529)